include/ignition/rendering/base/BaseJointVisual.hh

```diff
@@ -0,0 +1,548 @@
+          const std::string &_parentName)
+    {
+      this->parentAxis = _axis;
+      this->parentXyzExpressedIn = _xyzExpressedIn;
+      this->jointParentName = _parentName;
+      this->dirtyParentAxis = true;
+    }
+
+    /////////////////////////////////////////////////
+    template <class T>
+    void BaseJointVisual<T>::CreateParentAxis()
+    {
+      auto jointParentVis = this->Scene()->NodeByName(this->jointParentName);
+      if (jointParentVis == nullptr)
+      {
+        ignlog << "Joint parent with name " << this->jointParentName
+               << " does not exist"
+               << " so the parent axis will not be shown\n";
+        return;
+      }
+
+      if (this->parentAxisVis)
+      {
+        this->parentAxisVis->Destroy();
+        this->parentAxisVis.reset();
+      }
+
+      this->parentAxisVis = this->Scene()->CreateJointVisual();
+      jointParentVis->AddChild(this->parentAxisVis);
+      this->parentAxisVis->SetType(this->Type());
+      this->parentAxisVis->SetAxis(this->parentAxis,
+          this->parentXyzExpressedIn);
+
+      this->updateParentAxis = true;
+      this->ScaleToChild();
+    }
+
+    //////////////////////////////////////////////////
+    template <class T>
+    bool BaseJointVisual<T>::UpdateAxis(const ignition::math::Vector3d &_axis,
+          const std::string &_xyzExpressedIn)
+    {
+      if (this->ArrowVisual() && this->HasParent())
+      {
+        this->UpdateAxisImpl(this->ArrowVisual(), _axis, _xyzExpressedIn);
+        return true;
+      }
+
+      return false;
+    }
+
+    //////////////////////////////////////////////////
+    template <class T>
+    bool BaseJointVisual<T>::UpdateParentAxis(
+          const ignition::math::Vector3d &_axis,
+          const std::string &_xyzExpressedIn)
+    {
+      if (this->ParentAxisVisual() &&
+          this->ParentAxisVisual()->ArrowVisual() &&
+          this->ParentAxisVisual()->HasParent())
+      {
+        this->UpdateAxisImpl(this->ParentAxisVisual()->ArrowVisual(),
+            _axis, _xyzExpressedIn);
+        return true;
+      }
+
+      return false;
+    }
+
+    //////////////////////////////////////////////////
+    template <class T>
+    void BaseJointVisual<T>::UpdateAxisImpl(ArrowVisualPtr _arrowVisual,
+          const ignition::math::Vector3d &_axis,
+          const std::string &_xyzExpressedIn)
+    {
+      // Get rotation to axis vector
+      ignition::math::Vector3d axisDir = _axis;
+      ignition::math::Vector3d u = axisDir.Normalize();
+      ignition::math::Vector3d v = ignition::math::Vector3d::UnitZ;
+      double cosTheta = v.Dot(u);
+      double angle = acos(cosTheta);
+      ignition::math::Quaterniond quat;
+      // check the parallel case
+      if (ignition::math::equal(angle, IGN_PI))
+        quat.Axis(u.Perpendicular(), angle);
+      else
+        quat.Axis((v.Cross(u)).Normalize(), angle);
+      _arrowVisual->SetLocalRotation(quat);
+
+      if (_xyzExpressedIn == "__model__")
```

Comment:
what do you think about creating here a enum, instead of this string ? It looks weird

Reply:
The value of `_xyzExpressedIn` is `__model__` whenever the joint axis is expressed in the parent frame. How about a boolean value, for example, `useParentFrame` that is set to true whenever the joint axis is expressed in the parent frame? Using `_xyzExpressedIn` as a string looks more verbose, however.

Reply:
@iche033 thoughts?

Reply:
I think using the string `_model_` can be tricky. We can have joints connecting two different models and also nested models. Ign-rendering also doesn't have the concept of a 'model' as it's just a scene tree with nodes in hierarchical structure.  The current implementation looks like it's relative to its parent so I think a bool `useParentFrame`  fits better.

An alternative option is to take the name or id of a node as an argument. The function then creates the axis relative to that node.

Reply:
friendly ping @atharva-18

Reply:
> I think using the string `_model_` can be tricky. We can have joints connecting two different models and also nested models. Ign-rendering also doesn't have the concept of a 'model' as it's just a scene tree with nodes in hierarchical structure. The current implementation looks like it's relative to its parent so I think a bool `useParentFrame` fits better.
> 
> An alternative option is to take the name or id of a node as an argument. The function then creates the axis relative to that node.

The axis is relative to the parent visual, so using `useParentFrame` would be a better option (The user would also have to fetch and pass the parent node id every time in the other case).

https://github.com/ignitionrobotics/ign-rendering/blob/4cde82345b3653cd3ee9f1907770daae2859626c/include/ignition/rendering/base/BaseJointVisual.hh#L399-L404

Reply:
alright sounds good. Let's go with `useParentFrame`

Reply:
Added in [`bfbb44b`](https://github.com/ignitionrobotics/ign-rendering/pull/366/commits/bfbb44b4a988b723d38a79a499e70528413ec6e5)